File: wasm/src/runtime/durable-content.ts
export class DurableContentError extends Error {
  constructor(message: string) { super(message); this.name = "DurableContentError" }
}

export type DurableSource =
  | { kind: "url"; url: string }
  | { kind: "base64"; data: string }
  | { kind: "file_id"; id: string; affinity: { provider_id: string; endpoint_id: string } }
  | { kind: "object"; handle: string; owner: string; payload_ref: string }

export type DurableContentBlock =
  | { type: "text"; text: string }
  | { type: "image" | "audio" | "video" | "file"; source: DurableSource; media_type?: string; provider_options?: Record<string, unknown> }

export interface DurableContent { blocks: DurableContentBlock[] }
export interface DurableToolResult extends DurableContent { call_id: string; is_error: boolean }

function obj(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new DurableContentError(`${label} must be an object`)
  return value as Record<string, unknown>
}
function keys(value: Record<string, unknown>, allowed: string[], label: string): void {
  for (const key of Object.keys(value)) if (!allowed.includes(key)) throw new DurableContentError(`${label} has unknown field ${key}`)
}
function string(value: unknown, label: string): string {
  if (typeof value !== "string" || !value) throw new DurableContentError(`${label} must be a non-empty string`)
  return value
}
function boolean(value: unknown, label: string): boolean {
  if (typeof value !== "boolean") throw new DurableContentError(`${label} must be a boolean`)
  return value
}
function base64(value: unknown, label: string): string {
  const data = string(value, label)
  if (data.length % 4 !== 0 || !/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(data)) throw new DurableContentError(`${label} is not valid base64`)
  return data
}
function source(value: unknown, label: string): DurableSource {
  const raw = obj(value, `${label} source`)
  const kind = raw.kind
  if (kind === "url") { keys(raw, ["kind", "url"], `${label} source`); return { kind, url: string(raw.url, `${label} URL`) } }
  if (kind === "base64") { keys(raw, ["kind", "data"], `${label} source`); return { kind, data: base64(raw.data, `${label} base64 data`) } }
  if (kind === "file_id") {
    keys(raw, ["kind", "id", "affinity"], `${label} source`)
    const a = obj(raw.affinity, `${label} affinity`); keys(a, ["provider_id", "endpoint_id"], `${label} affinity`)
    return { kind, id: string(raw.id, `${label} file id`), affinity: { provider_id: string(a.provider_id, `${label} affinity provider_id`), endpoint_id: string(a.endpoint_id, `${label} affinity endpoint_id`) } }
  }
  if (kind === "object") {
    keys(raw, ["kind", "handle", "owner", "payload_ref"], `${label} source`)
    return { kind, handle: string(raw.handle, `${label} object handle`), owner: string(raw.owner, `${label} object owner`), payload_ref: string(raw.payload_ref, `${label} payload_ref`) }
  }
  throw new DurableContentError(`${label} source kind is invalid`)
}
function block(value: unknown, index: number): DurableContentBlock {
  const raw = obj(value, `content block ${index}`)
  if (raw.type === "text") { keys(raw, ["type", "text"], `content block ${index}`); return { type: "text", text: string(raw.text, `content block ${index} text`) } }
  if (raw.type === "tool_result") throw new DurableContentError("nested tool_result blocks are forbidden")
  if (raw.type !== "image" && raw.type !== "audio" && raw.type !== "video" && raw.type !== "file") throw new DurableContentError(`unknown content block type: ${String(raw.type)}`)
  keys(raw, ["type", "source", "media_type", "provider_options"], `content block ${index}`)
  const result: DurableContentBlock = { type: raw.type, source: source(raw.source, `content block ${index}`) }
  if (raw.media_type !== undefined) result.media_type = string(raw.media_type, `content block ${index} media_type`)
  if (raw.provider_options !== undefined) result.provider_options = obj(raw.provider_options, `content block ${index} provider_options`)
  return result
}
function blocks(value: unknown): DurableContentBlock[] { if (!Array.isArray(value)) throw new DurableContentError("content blocks must be an array"); return value.map(block) }

export function decodeDurableContent(value: unknown): DurableContent {
  const raw = obj(value, "durable content"); keys(raw, ["blocks"], "durable content")
  return { blocks: blocks(raw.blocks) }
}
export function decodeDurableToolResult(value: unknown): DurableToolResult {
  const raw = obj(value, "durable tool result")
  keys(raw, ["call_id", "is_error", "blocks"], "durable tool result")
  return { call_id: string(raw.call_id, "tool result call_id"), is_error: boolean(raw.is_error, "tool result is_error"), blocks: blocks(raw.blocks) }
}
export function encodeDurableContent(content: DurableContent): Record<string, unknown> { return decodeDurableContent(content) as unknown as Record<string, unknown> }
export function encodeDurableToolResult(result: DurableToolResult): Record<string, unknown> { return decodeDurableToolResult(result) as unknown as Record<string, unknown> }
export function toolOutputBlocksToDurable(blocks: readonly import("../types.js").ToolOutputBlock[]): DurableContentBlock[] {
  return blocks.map((part, index) => {
    if (part.type === "text") return { type: "text", text: part.text }
    const source = part.source
    const durableSource = source.kind === "fileId"
      ? {
          kind: "file_id" as const,
          id: source.id,
          affinity: source.affinity
            ? { provider_id: source.affinity.providerId, endpoint_id: source.affinity.endpointId }
            : (() => { throw new DurableContentError(`tool output block ${index} fileId source requires endpoint affinity`) })(),
        }
      : source.kind === "base64" ? { kind: "base64" as const, data: source.data }
      : source.kind === "url" ? { kind: "url" as const, url: source.url }
      : {
          kind: "object" as const,
          handle: source.handle,
          owner: source.owner ?? (() => { throw new DurableContentError(`tool output block ${index} object source requires owner`) })(),
          payload_ref: source.payloadRef ?? (() => { throw new DurableContentError(`tool output block ${index} object source requires payloadRef`) })(),
        }
    return { type: part.type, source: durableSource, ...(part.mediaType ? { media_type: part.mediaType } : {}), ...(part.providerOptions ? { provider_options: part.providerOptions } : {}) }
  })
}
export function durableBlocksToToolOutput(blocks: readonly DurableContentBlock[]): import("../types.js").ToolOutputBlock[] {
  return blocks.map((part) => {
    if (part.type === "text") return part
    const source = part.source
    const sdkSource = source.kind === "file_id"
      ? { kind: "fileId" as const, id: source.id, affinity: { providerId: source.affinity.provider_id, endpointId: source.affinity.endpoint_id } }
      : source.kind === "base64" ? source : source.kind === "url" ? source : { kind: "object" as const, handle: source.handle, owner: source.owner, payloadRef: source.payload_ref }
    return { type: part.type, source: sdkSource, ...(part.media_type ? { mediaType: part.media_type } : {}), ...(part.provider_options ? { providerOptions: part.provider_options } : {}) } as import("../types.js").ToolOutputBlock
  })
}

File: python/deepstrike/runtime/durable_content.py
"""Strict versioned durable content ABI (016-02).

This module is provider-neutral. It validates durable records before provider-specific projection;
legacy text and ``output``-only tool results are migrated to one text block.
"""
from __future__ import annotations

import base64
import binascii
from typing import Any


class DurableContentError(ValueError):
  pass


def _obj(value: Any, label: str) -> dict[str, Any]:
  if not isinstance(value, dict):
    raise DurableContentError(f"{label} must be an object")
  return value


def _keys(value: dict[str, Any], allowed: set[str], label: str) -> None:
  unknown = set(value) - allowed
  if unknown:
    raise DurableContentError(f"{label} has unknown field {sorted(unknown)[0]}")


def _string(value: Any, label: str) -> str:
  if not isinstance(value, str) or not value:
    raise DurableContentError(f"{label} must be a non-empty string")
  return value


def _boolean(value: Any, label: str) -> bool:
  if not isinstance(value, bool):
    raise DurableContentError(f"{label} must be a boolean")
  return value


def _source(value: Any, label: str) -> dict[str, Any]:
  raw = _obj(value, f"{label} source")
  kind = raw.get("kind")
  allowed = {
    "url": {"kind", "url"},
    "base64": {"kind", "data"},
    "file_id": {"kind", "id", "affinity"},
    "object": {"kind", "handle", "owner", "payload_ref"},
  }.get(kind)
  if allowed is None:
    raise DurableContentError(f"{label} source kind is invalid")
  _keys(raw, allowed, f"{label} source")
  if kind == "url":
    return {"kind": kind, "url": _string(raw.get("url"), f"{label} URL")}
  if kind == "base64":
    data = _string(raw.get("data"), f"{label} base64 data")
    try:
      base64.b64decode(data, validate=True)
    except (binascii.Error, ValueError) as exc:
      raise DurableContentError(f"{label} base64 data is not valid base64") from exc
    return {"kind": kind, "data": data}
  if kind == "file_id":
    out = {"kind": kind, "id": _string(raw.get("id"), f"{label} file id")}
    affinity = _obj(raw.get("affinity"), f"{label} affinity")
    _keys(affinity, {"provider_id", "endpoint_id"}, f"{label} affinity")
    out["affinity"] = {
      "provider_id": _string(affinity.get("provider_id"), f"{label} affinity provider_id"),
      "endpoint_id": _string(affinity.get("endpoint_id"), f"{label} affinity endpoint_id"),
    }
    return out
  return {
    "kind": kind,
    "handle": _string(raw.get("handle"), f"{label} object handle"),
    "owner": _string(raw.get("owner"), f"{label} object owner"),
    "payload_ref": _string(raw.get("payload_ref"), f"{label} payload_ref"),
  }


def _block(value: Any, index: int) -> dict[str, Any]:
  raw = _obj(value, f"content block {index}")
  kind = raw.get("type")
  if kind == "text":
    _keys(raw, {"type", "text"}, f"content block {index}")
    if not isinstance(raw.get("text"), str):
      raise DurableContentError(f"content block {index} text must be a string")
    return {"type": kind, "text": raw["text"]}
  if kind == "tool_result":
    raise DurableContentError("nested tool_result blocks are forbidden")
  if kind not in {"image", "audio", "video", "file"}:
    raise DurableContentError(f"unknown content block type: {kind}")
  _keys(raw, {"type", "source", "media_type", "provider_options"}, f"content block {index}")
  out = {"type": kind, "source": _source(raw.get("source"), f"content block {index}")}
  if "media_type" in raw:
    out["media_type"] = _string(raw["media_type"], f"content block {index} media_type")
  if "provider_options" in raw:
    out["provider_options"] = _obj(raw["provider_options"], f"content block {index} provider_options")
  return out


def _blocks(value: Any) -> list[dict[str, Any]]:
  if not isinstance(value, list):
    raise DurableContentError("content blocks must be an array")
  return [_block(item, index) for index, item in enumerate(value)]


def decode_durable_content(value: Any) -> dict[str, Any]:
  raw = _obj(value, "durable content")
  _keys(raw, {"schema_version", "blocks"}, "durable content")
  if raw.get("schema_version") != 1:
    raise DurableContentError(f"unsupported durable content schema_version {raw.get('schema_version')}")
  return {"schema_version": 1, "blocks": _blocks(raw.get("blocks"))}


def decode_durable_tool_result(value: Any) -> dict[str, Any]:
  raw = _obj(value, "durable tool result")
  if "blocks" not in raw and "output" in raw:
    _keys(raw, {"call_id", "output", "is_error"}, "legacy tool result")
    return {
      "schema_version": 1,
      "call_id": _string(raw.get("call_id"), "tool result call_id"),
      "is_error": _boolean(raw.get("is_error", False), "legacy tool result is_error"),
      "blocks": [{"type": "text", "text": _string(raw.get("output"), "legacy tool result output") if raw.get("output") != "" else ""}],
    }
  _keys(raw, {"schema_version", "call_id", "is_error", "blocks"}, "durable tool result")
  if raw.get("schema_version") != 1:
    raise DurableContentError(f"unsupported durable tool result schema_version {raw.get('schema_version')}")
  return {
    "schema_version": 1,
    "call_id": _string(raw.get("call_id"), "tool result call_id"),
    "is_error": _boolean(raw.get("is_error"), "tool result is_error"),
    "blocks": _blocks(raw.get("blocks")),
  }


def migrate_legacy_content(text: str) -> dict[str, Any]:
  if not isinstance(text, str):
    raise DurableContentError("legacy content must be a string")
  return {"schema_version": 1, "blocks": [{"type": "text", "text": text}]}


def encode_durable_content(content: dict[str, Any]) -> dict[str, Any]:
  return decode_durable_content(content)


def encode_durable_tool_result(result: dict[str, Any]) -> dict[str, Any]:
  return decode_durable_tool_result(result)


def durable_blocks_to_runtime(blocks: list[dict[str, Any]]) -> list[dict[str, Any]]:
  out: list[dict[str, Any]] = []
  for block in blocks:
    if block["type"] == "text":
      out.append(block)
      continue
    source = dict(block["source"])
    if source["kind"] == "file_id":
      source["kind"] = "fileId"
      if source.get("affinity"):
        source["affinity"] = {
          "providerId": source["affinity"]["provider_id"],
          "endpointId": source["affinity"]["endpoint_id"],
        }
    elif source["kind"] == "object":
      source["payloadRef"] = source.pop("payload_ref")
    out.append({"type": block["type"], "source": source,
      **({"media_type": block["media_type"]} if "media_type" in block else {}),
      **({"provider_options": block["provider_options"]} if "provider_options" in block else {})})
  return out


def runtime_blocks_to_durable(blocks: list[dict[str, Any]] | tuple[dict, ...]) -> list[dict[str, Any]]:
  """Freeze provider-carrier blocks with the durable ABI's snake_case source contract."""
  durable: list[dict[str, Any]] = []
  for block in blocks:
    kind = block.get("type")
    if kind == "text":
      durable.append({"type": "text", "text": block.get("text")})
      continue
    source = dict(block.get("source") or {})
    if source.get("kind") == "fileId":
      source["kind"] = "file_id"
      if source.get("affinity"):
        source["affinity"] = {
          "provider_id": source["affinity"].get("providerId"),
          "endpoint_id": source["affinity"].get("endpointId"),
        }
      else:
        raise DurableContentError("runtime fileId source requires endpoint affinity")
    elif source.get("kind") == "object":
      source["payload_ref"] = source.pop("payloadRef", source.get("payload_ref"))
    if source.get("kind") == "object" and (not source.get("owner") or not source.get("payload_ref")):
      raise DurableContentError("runtime object source requires owner and payloadRef")
    durable.append({
      "type": kind, "source": source,
      **({"media_type": block.get("media_type", block.get("mediaType"))} if block.get("media_type", block.get("mediaType")) else {}),
      **({"provider_options": block.get("provider_options", block.get("providerOptions"))} if block.get("provider_options", block.get("providerOptions")) else {}),
    })
  return decode_durable_content({"schema_version": 1, "blocks": durable})["blocks"]
